File: frontend/src/types/dividend.ts
export type DividendType = 'CASH' | 'STOCK' | 'DRIP'

export interface Dividend {
  id: string
  ticker: string
  stock_name: string | null
  dividend_type: DividendType
  amount: string       // Decimal from backend
  currency: string
  shares_received: string | null
  ex_dividend_date: string  // YYYY-MM-DD
  payment_date: string | null
  note: string | null
  account_id: string
  account_name: string
  created_at: string
}

export interface DividendCreate {
  ticker: string
  dividend_type: DividendType
  amount: string
  currency: string
  shares_received?: string | null
  ex_dividend_date: string
  payment_date?: string | null
  note?: string | null
  account_id: string
}

export interface DividendUpdate {
  dividend_type?: DividendType
  amount?: string
  currency?: string
  shares_received?: string | null
  ex_dividend_date?: string
  payment_date?: string | null
  note?: string | null
  account_id?: string | null
}

File: backend/app/models/dividend.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from sqlalchemy import Date, DateTime, ForeignKey, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.models.base import Base


class Dividend(Base):
    __tablename__ = "dividends"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    stock_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("stocks.id"), nullable=False)
    amount: Mapped[Decimal] = mapped_column(Numeric(18, 6), nullable=False)
    currency: Mapped[str] = mapped_column(String(5), nullable=False)
    ex_dividend_date: Mapped[date] = mapped_column(Date, nullable=False)
    payment_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    note: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )

    user: Mapped["User"] = relationship("User", back_populates="dividends")  # noqa: F821
    stock: Mapped["Stock"] = relationship("Stock", back_populates="dividends")  # noqa: F821
